{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"gpu","dataSources":[{"sourceId":8715559,"sourceType":"datasetVersion","datasetId":5229005},{"sourceId":8720499,"sourceType":"datasetVersion","datasetId":5232712},{"sourceId":8720551,"sourceType":"datasetVersion","datasetId":5232753},{"sourceId":8724638,"sourceType":"datasetVersion","datasetId":5235806}],"dockerImageVersionId":30177,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":true}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Is it Koyo? ","metadata":{}},{"cell_type":"markdown","source":"June 18, 2024 \n\nBy Natalie Ho \n\nAfter working on fast.ai's first module on classifying birds vs. dogs, I decided to see if I could make a module of my own identifying my friends, Koyo and Mitchell! ","metadata":{}},{"cell_type":"code","source":"#NB: Kaggle requires phone verification to use the internet or a GPU. If you haven't done that yet, the cell below will fail\n#    This code is only here to check that your internet is enabled. It doesn't do anything else.\n#    Here's a help thread on getting your phone number verified: https://www.kaggle.com/product-feedback/135367\n\nimport socket,warnings\ntry:\n    socket.setdefaulttimeout(1)\n    socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect(('1.1.1.1', 53))\nexcept socket.error as ex: raise Exception(\"STOP: No internet. Click '>|' in top right and set 'Internet' switch to on\")","metadata":{"execution":{"iopub.status.busy":"2024-06-19T00:54:24.324481Z","iopub.execute_input":"2024-06-19T00:54:24.325099Z","iopub.status.idle":"2024-06-19T00:54:24.336973Z","shell.execute_reply.started":"2024-06-19T00:54:24.325058Z","shell.execute_reply":"2024-06-19T00:54:24.336363Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# It's a good idea to ensure you're running the latest version of any libraries you need.\n# `!pip install -Uqq <libraries>` upgrades to the latest version of <libraries>\n# NB: You can safely ignore any warnings or errors pip spits out about running as root or incompatibilities\nimport os\niskaggle = os.environ.get('KAGGLE_KERNEL_RUN_TYPE', '')\n\nif iskaggle:\n    !pip install -U duckduckgo_search\n    !pip install fastai\n    !pip install fastbook","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.status.busy":"2024-06-19T00:54:27.000502Z","iopub.execute_input":"2024-06-19T00:54:27.001084Z","iopub.status.idle":"2024-06-19T00:55:10.852834Z","shell.execute_reply.started":"2024-06-19T00:54:27.001045Z","shell.execute_reply":"2024-06-19T00:55:10.852062Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 1: Download images of Koyo","metadata":{}},{"cell_type":"code","source":"from fastbook import *\nfrom fastai.vision.widgets import *\nfrom fastai.vision.all import *","metadata":{"_kg_hide-input":true,"execution":{"iopub.status.busy":"2024-06-19T00:55:10.854643Z","iopub.execute_input":"2024-06-19T00:55:10.854898Z","iopub.status.idle":"2024-06-19T00:55:13.693175Z","shell.execute_reply.started":"2024-06-19T00:55:10.854870Z","shell.execute_reply":"2024-06-19T00:55:13.692540Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"'''\nfrom fastdownload import download_url\ndest = 'bird.jpg'\ndownload_url(urls[0], dest, show_progress=False)\n\n\nim = Image.open(dest)\nim.to_thumb(256,256)\n'''","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Opening a picture of Koyo using Pillow's Image class. ","metadata":{}},{"cell_type":"code","source":"import pandas as pd\nfrom PIL import Image\n\n# Example of opening an image file\nimage_file_path = '/kaggle/input/new-images/Images/Koyo/IMG_4218 Small.jpeg'\nimage = Image.open(image_file_path).to_thumb(256,256)\nimage","metadata":{"execution":{"iopub.status.busy":"2024-06-19T01:47:18.205940Z","iopub.execute_input":"2024-06-19T01:47:18.206645Z","iopub.status.idle":"2024-06-19T01:47:18.251875Z","shell.execute_reply.started":"2024-06-19T01:47:18.206590Z","shell.execute_reply":"2024-06-19T01:47:18.251100Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 2: Train our model","metadata":{}},{"cell_type":"markdown","source":"Used fast.ai's DataBlock to randomly select 80% of images to be the training set and 20% to be the validation set. This function shows a sample of 9 images that classifies if an image contains Mitchell or Koyo in it. \n\n**Challenges**\n* Ran into an UnidentifiedImageError multiple times. I realized that I had to convert some images from HEIC into JPG. But even that wasn't enough! I had to compress all the images to avoid this error. \n* At first I had uploaded two separate folders of images. I realized I had to put the two Koyo and Mitchell subfolders into one larger Images folder as a complete dataset, so that DataBlock could be called and sort images of different categories. ","metadata":{}},{"cell_type":"code","source":"dls = DataBlock(\n    blocks=(ImageBlock, CategoryBlock), \n    get_items=get_image_files, \n    splitter=RandomSplitter(valid_pct=0.2, seed=42),\n    get_y=parent_label,\n    item_tfms=[Resize(192, method='squish')]\n).dataloaders('/kaggle/input/new-images/Images/', bs=40)\n#print(get_items)\ndls.show_batch(max_n=9)\n","metadata":{"execution":{"iopub.status.busy":"2024-06-19T01:00:49.542142Z","iopub.execute_input":"2024-06-19T01:00:49.542916Z","iopub.status.idle":"2024-06-19T01:00:52.149641Z","shell.execute_reply.started":"2024-06-19T01:00:49.542875Z","shell.execute_reply":"2024-06-19T01:00:52.148931Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"I thought one of the ways to fix the Unidentified image error was because the image was truncated, but turns out it was a different problem. \n\nHere I am fine tuning my model with resnet18, an image classification set pre-trained on ImageNet dataset. ","metadata":{}},{"cell_type":"code","source":"!pip install pillow\n\nfrom PIL import ImageFile\n\n#ImageFile.LOAD_TRUNCATED_IMAGES = True\n\n\nlearn = vision_learner(dls, resnet18, metrics=error_rate)\nlearn.fine_tune(3)","metadata":{"execution":{"iopub.status.busy":"2024-06-19T01:54:36.172953Z","iopub.execute_input":"2024-06-19T01:54:36.173618Z","iopub.status.idle":"2024-06-19T01:54:57.629654Z","shell.execute_reply.started":"2024-06-19T01:54:36.173574Z","shell.execute_reply":"2024-06-19T01:54:57.628843Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Step 3: Test the model","metadata":{}},{"cell_type":"code","source":"test = '/kaggle/input/new-images/Images/Koyo/IMG_4594 Small.jpeg'\nis_Koyo,_,probs = learn.predict(PILImage.create(test))","metadata":{"execution":{"iopub.status.busy":"2024-06-19T02:03:24.198687Z","iopub.execute_input":"2024-06-19T02:03:24.199363Z","iopub.status.idle":"2024-06-19T02:03:24.240684Z","shell.execute_reply.started":"2024-06-19T02:03:24.199321Z","shell.execute_reply":"2024-06-19T02:03:24.240086Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(f\"This is a: {is_Koyo}.\")\nprint(f\"Probability it's a Koyo: {probs[0]:.4f}\")\nimage = Image.open(test).to_thumb(256,256)\nimage","metadata":{"execution":{"iopub.status.busy":"2024-06-19T02:03:25.609079Z","iopub.execute_input":"2024-06-19T02:03:25.609367Z","iopub.status.idle":"2024-06-19T02:03:25.642072Z","shell.execute_reply.started":"2024-06-19T02:03:25.609333Z","shell.execute_reply":"2024-06-19T02:03:25.641422Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}